File: ServerCore/src/main/java/org/elkoserver/foundation/server/ManagerClassConnectionSetup.java
package org.elkoserver.foundation.server;

import org.elkoserver.foundation.net.MessageHandlerFactory;
import org.elkoserver.foundation.net.NetAddr;
import org.elkoserver.foundation.net.NetworkManager;
import org.elkoserver.foundation.properties.ElkoProperties;
import org.elkoserver.foundation.server.metadata.AuthDesc;
import org.elkoserver.util.trace.Trace;
import org.elkoserver.util.trace.TraceFactory;

import java.io.IOException;

class ManagerClassConnectionSetup extends BaseConnectionSetup {
    private String hostIncludingPortNumber;
    private final String mgrClass;
    private final NetworkManager myNetworkManager;
    private final MessageHandlerFactory actorFactory;

    ManagerClassConnectionSetup(String label, String mgrClass, String host, AuthDesc auth, boolean secure, ElkoProperties props, String propRoot, NetworkManager networkManager, MessageHandlerFactory actorFactory, Trace trServer, Trace tr, TraceFactory traceFactory) {
        super(label, host, auth, secure, props, propRoot, trServer, tr, traceFactory);

        this.mgrClass = mgrClass;
        hostIncludingPortNumber = host;
        myNetworkManager = networkManager;
        this.actorFactory = actorFactory;
    }

    @Override
    public String getProtocol() {
        return "manager class: " + mgrClass;
    }

    @Override
    public String getServerAddress() {
        return hostIncludingPortNumber;
    }

    @Override
    NetAddr tryToStartListener() throws IOException {
        NetAddr result = myNetworkManager.listenVia(
                mgrClass,
                propRoot,
                bind,
                actorFactory,
                msgTrace,
                secure);
        if (host.indexOf(':') < 0) {
            hostIncludingPortNumber = host + ":" + result.getPort();
        }
        return result;
    }

    @Override
    String getListenAddressDescription() {
        return hostIncludingPortNumber;
    }

    @Override
    String getValueToCompareWithBind() {
        return hostIncludingPortNumber;
    }

    @Override
    String getConnectionsSuffixForNotice() {
        return " using " + mgrClass;
    }

    @Override
    String getProtocolSuffixForErrorMessage() {
        return " (" + mgrClass + ")";
    }
}

File: ServerCore/src/main/kotlin/org/elkoserver/foundation/server/ManagerClassConnectionSetup.kt
package org.elkoserver.foundation.server

import org.elkoserver.foundation.net.MessageHandlerFactory
import org.elkoserver.foundation.net.NetAddr
import org.elkoserver.foundation.net.NetworkManager
import org.elkoserver.foundation.properties.ElkoProperties
import org.elkoserver.foundation.server.metadata.AuthDesc
import org.elkoserver.util.trace.TraceFactory
import org.elkoserver.util.trace.slf4j.Gorgel
import java.io.IOException

internal class ManagerClassConnectionSetup(
        label: String?,
        private val mgrClass: String,
        override var serverAddress: String,
        auth: AuthDesc,
        secure: Boolean,
        props: ElkoProperties,
        propRoot: String,
        private val myNetworkManager: NetworkManager,
        private val actorFactory: MessageHandlerFactory,
        gorgel: Gorgel,
        traceFactory: TraceFactory)
    : BaseConnectionSetup(label, serverAddress, auth, secure, props, propRoot, gorgel, traceFactory) {
    override val protocol: String
        get() = "manager class: $mgrClass"

    @Throws(IOException::class)
    override fun tryToStartListener(): NetAddr {
        val result = myNetworkManager.listenVia(
                mgrClass,
                propRoot,
                bind,
                actorFactory,
                msgTrace,
                secure)
        if (serverAddress.indexOf(':') < 0) {
            serverAddress = "$serverAddress:${result.port}"
        }
        return result
    }

    override val listenAddressDescription: String
        get() = serverAddress

    override val valueToCompareWithBind: String
        get() = serverAddress

    override val connectionsSuffixForNotice: String
        get() = " using $mgrClass"

    override val protocolSuffixForErrorMessage: String
        get() = " ($mgrClass)"
}
